llvm/lib/Target/AMDGPU/AMDGPUSwLowerLDS.cpp

```diff
@@ -0,0 +1,1335 @@
+        MetadataStructType, SwLDSMetadata, {Index0, Index1, Index2AlignedSize});
+
+    Value *AlignedDynLDSSize =
+        IRB.CreateAdd(CurrDynLDSSize, MaxAlignValueMinusOne);
+    AlignedDynLDSSize = IRB.CreateUDiv(AlignedDynLDSSize, MaxAlignValue);
+    AlignedDynLDSSize = IRB.CreateMul(AlignedDynLDSSize, MaxAlignValue);
+    IRB.CreateStore(AlignedDynLDSSize, GEPForAlignedSize);
+
+    // Update the Current Malloc Size
+    *CurrMallocSize = IRB.CreateAdd(*CurrMallocSize, AlignedDynLDSSize);
+  }
+}
+
+static DebugLoc getOrCreateDebugLoc(const Instruction *InsertBefore,
+                                    DISubprogram *SP) {
+  assert(InsertBefore);
+  if (InsertBefore->getDebugLoc())
+    return InsertBefore->getDebugLoc();
+  if (SP)
+    return DILocation::get(SP->getContext(), SP->getLine(), 1, SP);
+  return DebugLoc();
+}
+
+void AMDGPUSwLowerLDS::getLDSMemoryInstructions(
+    Function *Func, SetVector<Instruction *> &LDSInstructions) {
+  for (BasicBlock &BB : *Func) {
+    for (Instruction &Inst : BB) {
+      if (LoadInst *LI = dyn_cast<LoadInst>(&Inst)) {
+        if (LI->getPointerAddressSpace() == AMDGPUAS::LOCAL_ADDRESS)
+          LDSInstructions.insert(&Inst);
+      } else if (StoreInst *SI = dyn_cast<StoreInst>(&Inst)) {
+        if (SI->getPointerAddressSpace() == AMDGPUAS::LOCAL_ADDRESS)
+          LDSInstructions.insert(&Inst);
+      } else if (AtomicRMWInst *RMW = dyn_cast<AtomicRMWInst>(&Inst)) {
+        if (RMW->getPointerAddressSpace() == AMDGPUAS::LOCAL_ADDRESS)
+          LDSInstructions.insert(&Inst);
+      } else if (AtomicCmpXchgInst *XCHG = dyn_cast<AtomicCmpXchgInst>(&Inst)) {
+        if (XCHG->getPointerAddressSpace() == AMDGPUAS::LOCAL_ADDRESS)
+          LDSInstructions.insert(&Inst);
+      } else
+        continue;
+    }
+  }
+}
+
+Value *
+AMDGPUSwLowerLDS::getTranslatedGlobalMemoryGEPOfLDSPointer(Value *LoadMallocPtr,
+                                                           Value *LDSPtr) {
+  assert(LDSPtr && "Invalid LDS pointer operand");
+  Value *PtrToInt = IRB.CreatePtrToInt(LDSPtr, IRB.getInt32Ty());
+  Value *GEP =
+      IRB.CreateInBoundsGEP(IRB.getInt8Ty(), LoadMallocPtr, {PtrToInt});
+  return GEP;
+}
+
+void AMDGPUSwLowerLDS::translateLDSMemoryOperationsToGlobalMemory(
+    Function *Func, Value *LoadMallocPtr,
+    SetVector<Instruction *> &LDSInstructions) {
+  LLVM_DEBUG(dbgs() << "Translating LDS memory operations to global memory : "
+                    << Func->getName());
+  for (Instruction *Inst : LDSInstructions) {
+    IRB.SetInsertPoint(Inst);
+    if (LoadInst *LI = dyn_cast<LoadInst>(Inst)) {
+      Value *LIOperand = LI->getPointerOperand();
+      Value *Replacement =
+          getTranslatedGlobalMemoryGEPOfLDSPointer(LoadMallocPtr, LIOperand);
+      Value *NewLI = IRB.CreateAlignedLoad(LI->getType(), Replacement,
+                                           LI->getAlign(), LI->isVolatile());
+      AsanInfo.Instructions.insert(cast<Instruction>(NewLI));
```

Comment:
CreateAlignedLoad already returns LoadInst, make NewLI LoadInst and avoid the cast below